src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -131,154 +141,159 @@ protected override void DetachCore()
         {
             this.logger.LogTrace("()");
 
-            this.AttachedPeer.StateChanged -= this.AttachedNode_StateChanged;
+            this.AttachedPeer.MessageReceivedPriority -= this.AttachedPeer_MessageReceived;
+            this.AttachedPeer.StateChanged -= this.AttachedPeer_StateChanged;
 
             this.logger.LogTrace("(-)");
         }
 
-        private void Intercept(IncomingMessage message, Action continueInvocation)
+        private void AttachedPeer_MessageReceived(NetworkPeer peer, IncomingMessage message)
```

Comment:
I'm not a fan of such a big methods @Aprogiena. Can we not split the switch/case logic into separate methods here? Maybe we would want to write unit tests for the logic contained in the individual case statements? I would even split them off into classes but maybe that's a bit overkill. We need to try and stick to SOLID principles as much as we can and here we a method that is responsible for a lot of things.

WDYT?

Reply:
all this series of PRs is for network peer, it touched CHB here, so some work had to be done there, but further improvements of CHB are out of the scope of this PR

so for this PR, nope

otherwise, i have no problem with creating methods for each switch case as those are non trivial code blocks, so i agree. not into separate classes though, that's too much for me

but totally offtopic :)

Reply:
OK yeah, lets put those into methods, then at least we are one step closer to making the code more modular and testable 👍

Reply:
yes, but pls create a separate issue for that

Reply:
Is it going to be a lot of work to move the logic into methods?

Reply:
i don't know, it just does not belong here :-)
it is probably task of size 1, but still a separated task

Reply:
heheh ok cool man! Or maybe just add a TODO: so that we know to improve that area in another PR.

Reply:
ok i creted #856